File: src/Algorithm/Arrays/Leetcode Problems/Peak_Index_in_Mountain_Array/Peak_Index_in_Mountain_Array.cpp
class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
	    //start=1 and end=n-2 as peak can never be first and last element
        int start=1,end=arr.size()-2,answer;
        while(start<=end){

            int middle=start+(end-start)/2;
            //peak will always greater than left and right element
            if(arr[middle]>arr[middle-1] and arr[middle]>arr[middle+1]){
                answer=middle;
                break;
            }
			//if left element is greater than mid, peak will be in left side
            else if(arr[middle-1]>arr[middle])
                end=middle-1;
            else
                start=middle+1;
        }
        return answer;
    }
};

File: src/Algorithm/Arrays/LeetcodeProblems/Peak_Index_in_Mountain_Array/Peak_Index_in_Mountain_Array.java
package Algorithm.Arrays.LeetcodeProblems.Peak_Index_in_Mountain_Array;

public class Peak_Index_in_Mountain_Array {

    public static void main(String[] args){
      int[] arr = {0,1,0};
      System.out.println(peakIndexInMountainArray(arr));
    }
    public static int peakIndexInMountainArray(int[] arr) {
      int l = 0;
      int r = arr.length - 1;

      while (l < r) {
        final int m = (l + r) / 2;
        if (arr[m] < arr[m + 1])
          l = m + 1;
        else
          r = m;
      }

      return l;
    }
  }